### PR TITLE
Work fe1 ljankoschek linked org UI

### PR DESCRIPTION
**Pull request by ljankoschek**
UI Screens for the Federation Protocol with Tests. Added QRCode data structure and  Validator.

u1
ui2
ui3
ui4
ui5

### CODE REVIEW

File: fe1-web/src/core/network/validation/Validator.ts [hunk 1]
```diff
@@ -122,3 +124,12 @@ export function validateScannableMainPublicKey(obj: any): ValidationResult {
 export function validateScannablePopToken(obj: any): ValidationResult {
   return validate(`${schemaPrefix}/qrcode/pop_token.json`, obj);
 }
+
+export function validateFederationExchange(obj: any): ValidationResult {
+  // TODO change URL before merging to master
+  // also in federation_exchange.json
+  return validate(
+    `https://raw.githubusercontent.com/dedis/popstellar/work-fe1-ljankoschek-LinkedOrgUI/protocol/qrcode/federation_exchange.json`,
```
Comment (osm-alt) [inline fe1-web/src/core/network/validation/Validator.ts:132]: This can be changed to "https://raw.githubusercontent.com/dedis/popstellar/master/protocol/qrcode/federation_exchange.json" before merging as it's more of an identifier.

File: fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+                  onPress={() => {
+                    if (
+                      !manualLaoId ||
+                      !manualPublicKey ||
+                      !manualServerAddress ||
+                      !manualChallengeValue ||
+                      manualChallengeValidUntil <= Timestamp.EpochNow()
+                    ) {
+                      toast.show(
+                        `All fields are required and Valid Until has to be in the Future`,
```
[the remark this answers is not in the document]
Comment (ljankoschek) [reply fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx:385]: We have this problem currently in the whole application. I tried to fix this at the start of the semester but wasn't able to.

File: protocol/qrcode/federation_exchange.json [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            "type": "string",
+            "pattern": "^(ws|wss)://.*(:\\d{0,5})?/.*$",
+            "$comment": "public address of the remote organizer server"
+        },
+        "public_key": {
+            "type": "string",
+            "contentEncoding": "base64",
+            "$comment": "public key of the remote organizer"
+        },
+        "challenge": {
```
[the remark this answers is not in the document]
Comment (ljankoschek) [reply protocol/qrcode/federation_exchange.json:24]: Yes you are right! I changed it now so the challenge is not required.

File: fe1-web/src/core/network/validation/Validator.ts [hunk 4]
```diff
@@ -122,3 +124,12 @@ export function validateScannableMainPublicKey(obj: any): ValidationResult {
 export function validateScannablePopToken(obj: any): ValidationResult {
   return validate(`${schemaPrefix}/qrcode/pop_token.json`, obj);
 }
+
+export function validateFederationExchange(obj: any): ValidationResult {
+  // TODO change URL before merging to master
```
Comment (osm-alt) [inline fe1-web/src/core/network/validation/Validator.ts:129]: You can now remove this :)

**Review by osm-alt — APPROVED** [review]

File: fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+    top: '35%',
+  } as ViewStyle,
+  enterButton: {
+    ...QrCodeScannerUIElementContainer,
+    borderColor: Color.blue,
+    borderWidth: Spacing.x025,
+    alignSelf: 'center',
+    marginBottom: Spacing.contentSpacing,
+  } as ViewStyle,
+  flex1: {
```
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx:44]: If possible, I would try to give the classes meaningful names

File: fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+                    </Text>
+                  </PoPButton>
+                </View>
+                <PoPButton
+                  testID="show-scanner"
+                  buttonStyle="tertiary"
+                  onPress={() => {
+                    setShowQRScannerModal(!showQRScannerModal);
+                    setShowModal(!showModal);
+                    setShowScanner(!showScanner);
```
[the remark this answers is not in the document]
Comment (Tyratox) [reply fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx:205]: and actually in a few more places later

File: fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+                        valid_until: manualChallengeValidUntil,
+                      }),
+                    });
+                    onScanData(tmpOrg.toJson());
+                    setInputModalIsVisible(!inputModalIsVisible);
+                    setManualLaoID('');
+                    setManualPublicKey('');
+                    setManualServerAddress('');
+                    setManualChallengeValue('');
+                    setManualChallengeValidUntil(Timestamp.EpochNow());
```
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx:409]: If you set many variables at once, it may become more readable with useReducer? (https://react.dev/reference/react/useReducer)

File: fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+          <FAB
+            testID="fab-button"
+            placement="right"
+            color={accent}
+            onPress={() => {
+              setShowModal(true);
+            }}
+            icon={{ name: 'add', color: contrast }}
+            style={styles.fabStyle}
+          />
```
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx:425]: Are you sure you want to use a FAB here? This is not used anywhere else in the UI. iOS usually has these actions in the top navigation bar, top right corner. Right now fe1's UI tries to stick more towards the iOS side since there is a separate Android app :)

File: fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx:1]: This component is gigantic with an enormous amount of state. I would recommend to split this into multiple components :)

**Review by Tyratox — COMMENTED** [review]
Nice work! Here is a first batch of comments 😄

This PR does not include the network functionality yet, right?

**Comment by Tyratox** [comment]
Also I just saw that on the screenshots the text on the qr code and the FAB do not seem to be centered correctly?

File: fe1-web/src/core/network/validation/Validator.ts [hunk 9]
```diff
@@ -122,3 +124,10 @@ export function validateScannableMainPublicKey(obj: any): ValidationResult {
 export function validateScannablePopToken(obj: any): ValidationResult {
   return validate(`${schemaPrefix}/qrcode/pop_token.json`, obj);
 }
+
+export function validateFederationExchange(obj: any): ValidationResult {
+  return validate(
+    `https://raw.githubusercontent.com/dedis/popstellar/master/protocol/qrcode/federation_exchange.json`,
```
[the remark this answers is not in the document]
Comment (ljankoschek) [reply fe1-web/src/core/network/validation/Validator.ts:130]: I have only orientated myself here according to the other features, these also use the core/validator (e.g. lao feature with validateScannableMainPublicKey or home feature with validateConnectToLao).

File: fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+                    bottomDivider
+                    key={organization.lao_id.valueOf()}>
+                    <PoPIcon name="business" />
+                    <ListItem.Content>
+                      <ListItem.Title>Lao ID: {organization.lao_id}</ListItem.Title>
+                      <ListItem.Subtitle>
+                        Public Key: {organization.public_key}, Server Address:{' '}
+                        {organization.server_address}, Challenge Value:{' '}
+                        {organization.challenge.value}, Challenge Valid_until:{' '}
+                        {organization.challenge.valid_until.valueOf()}
```
[the remark this answers is not in the document]
Comment (ljankoschek) [reply fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx]: I only left the "Lao Id:" and put that into strings.ts. The other values I displayed only for debugging.

File: fe1-web/src/features/lao/navigation/LaoNavigation.tsx [hunk 11]
```diff
@@ -110,6 +110,20 @@ const LaoDrawerContent = ({ descriptors, navigation, state }: DrawerContentCompo
...
+    let result = iterator.next();
+    while (!result.done) {
+      const [index, value] = result.value;
+      if (value.id === 'Linked Organizations') {
+        passedScreens.splice(index, 1);
+      }
+      result = iterator.next();
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (ljankoschek) [reply fe1-web/src/features/lao/navigation/LaoNavigation.tsx:126]: To make it consistent with fe2, we decided that the page is visible for everyone, but only the organizer can use the add button.

File: fe1-web/src/core/components/QRCode.tsx
Comment (ljankoschek) [inline fe1-web/src/core/components/QRCode.tsx:1]: I had to change this, because otherwise the QRCodeof federation_exchange.json that doesn't include a challenge wasn't recognised as qrcode

File: fe1-web/src/features/home/screens/Launch.tsx
Comment (ljankoschek) [inline fe1-web/src/features/home/screens/Launch.tsx:1]: I am not sure whether or not this is the correct place to subscribe to the federation channel

File: fe1-web/src/features/linked-organizations/components/AddLinkedOrganizationModal.tsx [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+        console.error('Could not request Challenge, error:', err);
+        toast.show(`Could not request Challenge, error: ${err}`, {
+          type: 'danger',
+          placement: 'bottom',
+          duration: FOUR_SECONDS,
+        });
+      });
+  }, [laoId, toast]);
+
+  const getQRCodeData = () => {
```
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/components/AddLinkedOrganizationModal.tsx:89]: In general I would try to avoid prefixing functions with side effects (and not even returning a value) with `get` :) Is there a better name you can think of that may explain what the function actually does? (It definitely does not just get some value and return it, i.e. what you usually expect from a getX function)

File: fe1-web/src/features/linked-organizations/components/ManualInputModal.tsx [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+  isInitiatingOrganizer,
+  onScanData,
+}) => {
+  const toast = useToast();
+  const [manualLaoId, setManualLaoID] = useState<string>('');
+  const [manualPublicKey, setManualPublicKey] = useState<string>('');
+  const [manualServerAddress, setManualServerAddress] = useState<string>('');
+  const [manualChallengeValue, setManualChallengeValue] = useState<string>('');
+  const [manualChallengeValidUntil, setManualChallengeValidUntil] = useState<Timestamp>(
+    Timestamp.EpochNow().addSeconds(86400),
```
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/components/ManualInputModal.tsx:34]: Try to avoid magic numbers in the code. For time it already becomes more readable if you write `60 * 60 * 24` instead. Even better if you define constants in a central place saying something like `DAY_IN_SECONDS = 24 * HOUR_IN_SECONDS; HOUR_IN_SECONDS = 60 * MINUTE_IN_SECONDS; MINUTE_IN_SECONDS = 60`;

File: fe1-web/src/features/linked-organizations/components/ManualInputModal.tsx [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+            if (
+              (isInitiatingOrganizer &&
+                (!manualLaoId ||
+                  !manualPublicKey ||
+                  !manualServerAddress ||
+                  !manualChallengeValue ||
+                  manualChallengeValidUntil <= Timestamp.EpochNow())) ||
+              (!isInitiatingOrganizer && (!manualLaoId || !manualPublicKey || !manualServerAddress))
+            ) {
+              toast.show(`All fields are required and Valid Until has to be in the Future`, {
```
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/components/ManualInputModal.tsx:97]: another string for strings.ts? :)

File: fe1-web/src/features/linked-organizations/components/ManualInputModal.tsx [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+          testID="add-manually"
+          onPress={() => {
+            if (
+              (isInitiatingOrganizer &&
+                (!manualLaoId ||
+                  !manualPublicKey ||
+                  !manualServerAddress ||
+                  !manualChallengeValue ||
+                  manualChallengeValidUntil <= Timestamp.EpochNow())) ||
+              (!isInitiatingOrganizer && (!manualLaoId || !manualPublicKey || !manualServerAddress))
```
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/components/ManualInputModal.tsx:95]: With this simple logic it is okay-ish (I know we don't have proper form validation in other parts of the app) but the error message could be much more useful if it only showed relevant information such as error X with field Y.

Comment (Tyratox) [reply fe1-web/src/features/linked-organizations/components/ManualInputModal.tsx:95]: Maybe we should create an issue for proper form validation..

File: fe1-web/src/features/linked-organizations/components/QRCodeScannerModal.tsx [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+      transparent
+      visible={visible}
+      onRequestClose={onClose}
+      style={styles.modalContainer}>
+      <TouchableWithoutFeedback containerStyle={ModalStyles.modalBackground} onPress={onClose} />
+      <View style={styles.modalContainer}>
+        <ModalHeader onClose={onClose}>
+          {STRINGS.linked_organizations_addlinkedorg_scanQRCode}
+        </ModalHeader>
+        <Text style={{ ...Typography.paragraph, ...styles.textAlignCenter }}>
```
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/components/QRCodeScannerModal.tsx:66]: Not that anything is wrong with this but just wanted to let you know that you can also pass in an array of styles :)

File: fe1-web/src/features/linked-organizations/navigation/LinkedOrganizationsNavigation.tsx [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+          headerTitle: STRINGS.navigation_linked_organizations_title,
+          headerLeft: DrawerMenuButton,
+          headerRight: AddLinkedOrganizationButton,
+        }}
+      />
+      <LinkedOrganizationsNavigator.Screen
+        name={STRINGS.linked_organizations_navigation_addlinkedorgModal}
+        component={AddLinkedOrganizationModal}
+        options={{
+          presentation: 'transparentModal',
```
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/navigation/LinkedOrganizationsNavigation.tsx:35]: Huh, I didn't know about this :)

File: fe1-web/src/features/linked-organizations/network/LinkedOrgHandler.ts [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    const makeErr = (err: string) => `challenge/request was not processed: ${err}`;
+
+    const laoId = getCurrentLaoId();
+    if (!laoId) {
+      console.warn(makeErr('no Lao is currently active'));
+      return false;
+    }
+    if (msg.messageData instanceof ChallengeRequest) {
```
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/network/LinkedOrgHandler.ts:71]: Given we don't do this check in any other handler, why did you add it here? Could we add this to all other handlers instead / on top of the `as X` ?

File: fe1-web/src/features/linked-organizations/network/LinkedOrgHandler.ts [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+      const challengeRequest = msg.messageData as ChallengeRequest;
+      if (challengeRequest.timestamp) {
+        return true;
```
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/network/LinkedOrgHandler.ts:74]: How come we don't need to store this message anywhere?

File: fe1-web/src/features/linked-organizations/network/LinkedOrgHandler.ts [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+
+    if (msg.messageData instanceof FederationInit) {
+      const federationInit = msg.messageData as FederationInit;
+      if (
+        federationInit.lao_id &&
+        federationInit.public_key &&
+        federationInit.server_address &&
+        federationInit.challenge
+      ) {
+        return true;
```
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/network/LinkedOrgHandler.ts:110]: How come we don't need to store this message anywhere?

File: fe1-web/src/features/linked-organizations/network/LinkedOrgHandler.ts [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+
+    if (msg.messageData instanceof FederationExpect) {
+      const federationExpect = msg.messageData as FederationExpect;
+      if (
+        federationExpect.lao_id &&
+        federationExpect.public_key &&
+        federationExpect.server_address &&
+        federationExpect.challenge
+      ) {
+        return true;
```
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/network/LinkedOrgHandler.ts:146]: same questions as before

File: fe1-web/src/features/linked-organizations/network/messages/index.ts
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/network/messages/index.ts:1]: I guess you either forgot to remove this file or forgot to add the remaining messages?

File: fe1-web/src/features/linked-organizations/reducer/ChallengeReducer.ts [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+        return {
+          payload: {
+            laoId: laoId.valueOf(),
+            challenge: challenge,
+          },
+        };
+      },
+      reducer(state, action: PayloadAction<{ laoId: string; challenge: ChallengeState }>) {
+        const { laoId, challenge } = action.payload;
+        state.byLaoId[laoId] = challenge;
```
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/reducer/ChallengeReducer.ts:37]: Given this implementation it is more of a `setChallenge` rather than a `addChallenge`, no?

File: fe1-web/src/features/linked-organizations/reducer/__tests__/LinkedOrganizationsReducer.test.ts [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+              allLaoIds: [serializedMockLaoId],
+              byLinkedLaoId: { [serializedMockLaoId]: mockOrganizationState },
+            },
+          },
+        } as LinkedOrganizationReducerState,
+      }),
+    ).toEqual(mockOrganizationState);
+  });
+
+  it('returns undefined if the linked organization  is not in the store', () => {
```
Comment (Tyratox) [inline fe1-web/src/features/linked-organizations/reducer/__tests__/LinkedOrganizationsReducer.test.ts:109]:
```suggestion
  it('returns undefined if the linked organization is not in the store', () => {
```

File: fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx (hunk 5 above)
Comment (Tyratox) [reply fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx:15]: This was marked as resolved, I guess you don't agree? 😂

File: fe1-web/src/resources/strings.ts [hunk 24]
```diff
@@ -305,7 +311,7 @@ namespace STRINGS {
     'The Roll Call is currently open and you as the organizer should start adding attendees by scanning their PoP tokens.';
   export const roll_call_open_attendee =
     'The Roll Call is currently open and you as an attendee should let the organizer scan your PoP token encoded in the QR Code below.';
-  export const roll_call_qrcode_text = 'Scan to add attendee';
+  export const roll_call_qrcode_text = 'Scan to\nadd\nattendee';
```
Comment (Tyratox) [inline fe1-web/src/resources/strings.ts:314]: Are the newlines necessary? Did the automatic text wrapping break?

File: protocol/qrcode/federation_exchange.json [hunk 25]
```diff
@@ -31,7 +26,9 @@
             "properties": {
                 "value": {
                     "type": "string",
-                    "description": "The challenge value, which is a string of 32 random bytes encoded in base64 format"
+                    "contentEncoding": "hex",
+                    "pattern": "^[0-9a-fA-F]{64}$",
+                    "$comment": "A 32 bytes array encoded in hexadecimal"
```
Comment (Tyratox) [inline protocol/qrcode/federation_exchange.json:31]: Not that you need to change anything here, just curious: Why did you chose to go with hex instead of base64?

**Review by Tyratox — APPROVED** [review]
Really good work! Also super happy to see you adding so many tests!

I did add some comments but the PR looks quite clean overall :)

Quick question: I think I missed sth, how is the user on the UI informed that the federation succeeded?

File: fe1-web/src/resources/strings.ts (hunk 24 above)
Comment (ljankoschek) [reply fe1-web/src/resources/strings.ts:314]: For me it didn't work here in qr code overlay

File: fe1-web/src/features/linked-organizations/components/AddLinkedOrganizationButton.tsx [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+      (idx) => {
+        if (idx !== undefined && idx === 0) {
+          navigation.navigate(STRINGS.navigation_app_lao, {
+            screen: STRINGS.navigation_linked_organizations_title,
+            params: {
+              screen: STRINGS.linked_organizations_navigation_addlinkedorgModal,
+            },
+          });
+        } else {
+          // cancel
```
[the remark this answers is not in the document]
Comment (ljankoschek) [reply fe1-web/src/features/linked-organizations/components/AddLinkedOrganizationButton.tsx:42]: The cancel button was also in the other components with this type of action button that shows a sort of menu (events, notification,...)

File: protocol/qrcode/federation_exchange.json (hunk 25 above)
Comment (ljankoschek) [reply protocol/qrcode/federation_exchange.json:31]: Honestly, when we discussed the data authentication protocol we just described to use either base64 or hex and the backend guys started to implement it with hex..

File: fe1-web/src/features/linked-organizations/network/LinkedOrgHandler.ts (hunk 21 above)
Comment (ljankoschek) [reply fe1-web/src/features/linked-organizations/network/LinkedOrgHandler.ts:146]: Currently, as far as I understood it we don't need to store anything except the Challenge. We discussed in our latest meeting that we need to Broadcast the Federation Result Message, so that the Organizer knows the linking of the Organizations worked, and we only need to store that message then.  If it is fine by you I will do this in the next PR then, so I can merge this before the global meeting.

File: fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx (hunk 5 above)
Comment (ljankoschek) [reply fe1-web/src/features/linked-organizations/screens/LinkedOrganizationsScreen.tsx:15]: Sorry I overlooked that!

File: fe1-web/src/features/linked-organizations/network/LinkedOrgHandler.ts (hunk 18 above)
Comment (ljankoschek) [reply fe1-web/src/features/linked-organizations/network/LinkedOrgHandler.ts:71]: I think this also has something to do with that I don't save the ChallengeRequest. Most of the other handlers do the ... as XMessage and then do a try catch to create the Actual Object (e.g. RollCall) out of it. Since I am not creating an Object of the message I just checked it with the instanceOf and then checked for my variables (e.g. timestamp)

File: fe1-web/src/features/linked-organizations/components/ManualInputModal.tsx (hunk 15 above)
Comment (ljankoschek) [reply fe1-web/src/features/linked-organizations/components/ManualInputModal.tsx:95]: I changed it now so that it provides more precise error messages and I displayed that on the Modal as an error message and not in the toaster, because we still have the problem of the toaster being behind modals. I tried looking into that again but everything I tried did not work unfortunately.

Comment (Tyratox) [reply fe1-web/src/features/linked-organizations/components/ManualInputModal.tsx]: I see 😅 Great that you added better error messages though 😁

File: fe1-web/src/resources/strings.ts (hunk 24 above)
Comment (Tyratox) [reply fe1-web/src/resources/strings.ts:314]: Oh okay this is not good 😅 But I guess this is a good fix for now

File: protocol/qrcode/federation_exchange.json (hunk 25 above)
Comment (Tyratox) [reply protocol/qrcode/federation_exchange.json:31]: huh okay, interesting :)

**Review by Tyratox — APPROVED** [review]
